tsl/src/remote/connection_cache.c

```diff
@@ -156,49 +203,226 @@ connection_cache_create()
-{
-	return ts_cache_pin(connection_cache_current);
-}
-
 TSConnection *
-remote_connection_cache_get_connection(Cache *cache, TSConnectionId id)
+remote_connection_cache_get_connection(TSConnectionId id)
 {
 	CacheQuery query = { .data = &id };
-	ConnectionCacheEntry *entry = ts_cache_fetch(cache, &query);
+	ConnectionCacheEntry *entry = ts_cache_fetch(connection_cache, &query);
 
 	return entry->conn;
 }
 
-void
-remote_connection_cache_remove(Cache *cache, TSConnectionId id)
+bool
+remote_connection_cache_remove(TSConnectionId id)
 {
-	ts_cache_remove(cache, &id);
+	return ts_cache_remove(connection_cache, &id);
 }
 
 /*
  * Connection invalidation callback function
  *
  * After a change to a pg_foreign_server catalog entry,
- * mark the cache as invalid.
+ * mark the cache entry as invalid.
+ */
+void
+remote_connection_cache_invalidate_callback(Datum arg, int cacheid, uint32 hashvalue)
+{
+	HASH_SEQ_STATUS scan;
+	ConnectionCacheEntry *entry;
+
+	if (cacheid != FOREIGNSERVEROID)
+		return;
+
+	hash_seq_init(&scan, connection_cache->htab);
+
+	while ((entry = hash_seq_search(&scan)) != NULL)
+	{
+		/* hashvalue == 0 means cache reset, so invalidate entire cache */
+		if (hashvalue == 0 || entry->hashvalue == hashvalue)
+			entry->invalidated = true;
+	}
+}
+
+/*
+ * Check if a connection is local.
+ *
+ * This is an imperfect check, but should work for common cases.
+ *
+ * It currently doesn't capture being connected on a local network interface
+ * address. That would require a platform-independent way to lookup local
+ * interface addresses (on UNIX one could use getifaddrs, for instance).
+ */
+static bool
+is_local_connection(const PGconn *conn)
+{
+	const char *host = PQhost(conn);
+
+	return (strcmp("localhost", host) == 0 || strcmp("127.0.0.1", host) ||
+			host[0] == '/' /* unix domain socket starts with a slash */);
+}
+
+/*
+ * Remove connections that connect to a local DB, which is being dropped.
+ *
+ * This function is called whan a database is dropped on the local instance
+ * and is needed to prevent errors when the database serves as a
+ * "same-instance" data node (common in tests). When a data node exists on the
+ * local instance (i.e., in a separate local database), we need to close all
+ * connections to the data node database in order to drop it. Otherwise, the
+ * drop database will fail since the database is being used by the connections
+ * in the cache.
  */
 void
-remote_connection_cache_invalidate_callback(void)
+remote_connection_cache_dropped_db_callback(const char *dbname)
+{
+	HASH_SEQ_STATUS scan;
+	ConnectionCacheEntry *entry;
+
+	hash_seq_init(&scan, connection_cache->htab);
+
+	while ((entry = hash_seq_search(&scan)) != NULL)
+	{
+		PGconn *pgconn = remote_connection_get_pg_conn(entry->conn);
+
+		/* Remove the connection if is local and connects to a DB with the
+		 * same name as the one being dropped. */
+		if (strcmp(dbname, PQdb(pgconn)) == 0 && is_local_connection(pgconn))
```

Comment:
Maybe we should check port as well? Not sure if this case is possible here, what if we connect to some other local instance on different port but having the same db name

Reply:
Good point. I will add a check in is_local_connection

Reply:
dont you want to check for instance here and dont really care if its local or remote

Reply:
@svenklemm Not sure what you mean. Can you reformulate?

FWIW, this check is for dropping a database on the same instance as the connection connects to. If the connection is to a remote instance, then we don't care.

Reply:
Yes but the remote/local distinction is not important you only care if its the same instance even if its local

Reply:
@svenklemm do you have a suggestion how to check for "same instance"?

Reply:
could check it against information in MyPort https://doxygen.postgresql.org/structPort.html

Reply:
Still don't understand how that helps (code examples welcome).  
Conceptually, the connections look like this:

```
client -> backend_A -> backend_B
```
As backend_A, we need to identify if the connection from backend_A to backend_B is to the local instance, i.e., backend_A and backend_B were forked from the same postmaster

I can access `MyProcPort`, but that only gives the (current) backend A's connection info (i.e., the client). I guess if was possible to get access to backend B's Port struct via shared memory I could somehow compare that against the connection info.

Reply:
laddr has information about the server, i guess the most resilient way would be to check our own uuid which we could cache on connection setup, checking only against laddr might still fail in case postmaster is listening on multiple interfaces but it would be huge improvement over checking against "localhost" and "127.0.0.1"